### ISSUE TITLE
Database statistics not working properly - empty values?

### ISSUE THREAD

**Issue by endro** [taginfo/taginfo#266]
I have just created fresh taginfo instance on my new shared hosting but some reports not working properly. I.e. http://taginfo.openstreetmap.pl/reports/characters_in_keys show null/NaN values or http://taginfo.openstreetmap.pl/reports/database_statistics throwing error FloatDomainError with means trying convert NaN values to integer. Other reports working fine. I am browsing file taginfo-db.db using sqlite3 and I see that all values exists (nodes|133108866 nodes_without_tags|126447742 node_tags|26088805 etc.) Where is the problem?

**Comment by endro** [comment]
So historic_development report is working when I modify this line:
https://github.com/taginfo/taginfo/blob/cc1a7403488c802d5a59d3227b3dab05bab10225/web/viewsjs/reports/historic_development.js.erb#L6
to
`    data[key] = @db.execute("SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate",
key.force_encoding("UTF-8")).map do |row|`.

Original encoding of query string and bind_params is ASCII-8BIT. Taginfo-db.db database encoding is UTF-8.

But it is working too when I change it to:
`    data[key] = @db.execute("SELECT udate, value FROM history.history_stats WHERE key='"+key+"' ORDER BY udate").map do |row|`

What's going on?

**Comment by joto** [comment]
Did you run the import twice? "To create some statistics during an update, taginfo needs the statistics from the last update run. So you will have to run the update twice before you get all statistics to work properly. (But most of the stuff will already work after the first run.)" (from https://wiki.openstreetmap.org/wiki/Taginfo/Installation)

**Comment by endro** [comment]
Yes. This database has been updated four times. Before this I was doing every time clean update.

Original queries in log for report/historic_development: 
>SQL duration=0.0 query="SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate;" params=['num_keys']
>SQL duration=0.0 query="SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate;" params=['num_tags']
>SQL duration=0.0 query="SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate;" params=['relation_types']

It looks the same as queries with added force_encoding("UTF8") for @db.execute second parameter:
>SQL duration=0.0 query="SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate;" params=['num_keys']
>SQL duration=0.0 query="SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate;" params=['num_tags']
>SQL duration=0.0 query="SELECT udate, value FROM history.history_stats WHERE key=? ORDER BY udate;" params=['relation_types']

But first one return empty results set (i don't know ruby so i am debuging using logger.fatal):
>[26/Aug/2019:18:34:44 +0200] "GET /reports/historic_development HTTP/1.1" 200 5864 0.0155
>F, [2019-08-26T18:34:44.838385 #33635] FATAL -- : []
>F, [2019-08-26T18:34:44.838561 #33635] FATAL -- : []
>F, [2019-08-26T18:34:44.838734 #33635] FATAL -- : []`

And the second return all needed data:
> [26/Aug/2019:18:34:56 +0200] "GET /reports/historic_development HTTP/1.1" 200 5864 0.0157
>F, [2019-08-26T18:34:56.766508 #33635] FATAL -- : [{"udate"=>"2018-09-19", "value"=>6159, 0=>"2018-09-19", 1=>6159}, {"udate"=>"2018-09-19", "value"=>6159, 0=>"2018-09-19", 1=>6159}, {"udate"=>"2018-09-19", "value"=>6159, 0=>"2018-09-19", 1=>6159}, {"udate"=>"2019-04-13", "value"=>6391, 0=>"2019-04-13", 1=>6391}, {"udate"=>"2019-08-25", "value"=>6611, 0=>"2019-08-25", 1=>6611}]
>F, [2019-08-26T18:34:56.766743 #33635] FATAL -- : [{"udate"=>"2018-09-19", "value"=>1798666, 0=>"2018-09-19", 1=>1798666}, {"udate"=>"2018-09-19", "value"=>1798666, 0=>"2018-09-19", 1=>1798666}, {"udate"=>"2018-09-19", "value"=>1798666, 0=>"2018-09-19", 1=>1798666}, {"udate"=>"2019-04-13", "value"=>1835985, 0=>"2019-04-13", 1=>1835985}, {"udate"=>"2019-08-25", "value"=>1893475, 0=>"2019-08-25", 1=>1893475}]
>F, [2019-08-26T18:34:56.766952 #33635] FATAL -- : [{"udate"=>"2018-09-19", "value"=>81, 0=>"2018-09-19", 1=>81}, {"udate"=>"2018-09-19", "value"=>81, 0=>"2018-09-19", 1=>81}, {"udate"=>"2018-09-19", "value"=>81, 0=>"2018-09-19", 1=>81}, {"udate"=>"2019-04-13", "value"=>80, 0=>"2019-04-13", 1=>80}, {"udate"=>"2019-08-25", "value"=>78, 0=>"2019-08-25", 1=>78}]

**Comment by endro** [comment]
Why
>logger.fatal Encoding.default_external
>logger.fatal "newstring".encoding

Returns:
>#<Encoding:UTF-8>
>#<Encoding:ASCII-8BIT>

**Comment by endro** [comment]
I didn't find the cause of the problem, so I made some changes to the code. Should I create a new pull request?
[diff.txt](https://github.com/taginfo/taginfo/files/3580597/diff.txt)

**Comment by joto** [comment]
I am reluctant to accept a patch that we don't understand the reason for and that might have side-effects on other peoples installations. Nobody else has reported this problem, to I assume this has something to do with your particular Ruby setup or something like it. Maybe you can tell us about your setup and we can find something from that.

**Comment by joto** [comment]
I tracked a similar problem to the use of the `erubi` gem. Without that `gem` everything was fine. Maybe that will help you, too.
